lightning/src/ln/channelmanager.rs

```diff
@@ -4509,12 +4568,91 @@ impl_writeable_tlv_based!(HTLCPreviousHopData, {
+impl Writeable for ClaimableHTLC {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		let payment_data = match &self.claim_type {
+			ClaimableType::Invoice(data) => data.clone(),
+			_ => msgs::FinalOnionHopData {
+				payment_secret: PaymentSecret([0; 32]),
+				total_msat: 0,
+			}
+		};
+		let keysend_preimage = match self.claim_type {
+			ClaimableType::Invoice(_) => None,
+			ClaimableType::Spontaneous(preimage) => Some(preimage.clone()),
+		};
+		write_tlv_fields!
+		(writer,
+		 {
+		   (0, self.prev_hop, required), (2, self.value, required),
+		   (4, payment_data, required), (6, self.cltv_expiry, required),
+			 (8, keysend_preimage, option),
+		 });
+		Ok(())
+	}
+
+	#[inline]
+	fn serialized_length(&self) -> usize {
+		use util::ser::BigSize;
+		let len = {
+			#[allow(unused_mut)]
+			let mut len = ::util::ser::LengthCalculatingWriter(0);
+			get_varint_length_prefixed_tlv_length!(len, 0, self.prev_hop, required);
+			get_varint_length_prefixed_tlv_length!(len, 2, self.value, required);
+			let payment_data = match &self.claim_type {
+				ClaimableType::Invoice(data) => data.clone(),
+				_ => msgs::FinalOnionHopData {
+					payment_secret: PaymentSecret([0; 32]),
+					total_msat: 0,
+				}
+			};
+			get_varint_length_prefixed_tlv_length!(len, 4, payment_data, required);
+			get_varint_length_prefixed_tlv_length!(len, 6, self.cltv_expiry, required);
+			let keysend_preimage = match self.claim_type {
+				ClaimableType::Invoice(_) => None,
+				ClaimableType::Spontaneous(preimage) => Some(preimage.clone()),
+			};
+			get_varint_length_prefixed_tlv_length!(len, 8, keysend_preimage, option);
+			len.0
+		};
+		let mut len_calc = ::util::ser::LengthCalculatingWriter(0);
+		BigSize(len as u64).write(&mut len_calc).expect("No in-memory data may fail to serialize");
+		len + len_calc.0
+	}
+}
+
+impl Readable for ClaimableHTLC {
+	fn read<R: Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		init_tlv_field_var!(prev_hop, required);
+		init_tlv_field_var!(value, required);
+		init_tlv_field_var!(payment_data, required);
+		init_tlv_field_var!(cltv_expiry, required);
+		init_tlv_field_var!(keysend_preimage, option);
+		read_tlv_fields!
+		(reader,
+		 {
+		   (0, prev_hop, required), (2, value, required),
+		   (4, payment_data, required), (6, cltv_expiry, required),
+			 (8, keysend_preimage, option)
+		 });
+		let claim_type = match keysend_preimage {
+			Some(p) => {
+				let data: msgs::FinalOnionHopData = payment_data.0.unwrap();
+				if data.payment_secret != PaymentSecret([0; 32]) || data.total_msat != 0 {
+					return Err(DecodeError::InvalidValue)
+				}
+				ClaimableType::Spontaneous(p)
+			},
+			None => ClaimableType::Invoice(payment_data.0.unwrap()),
+		};
+		Ok(Self {
+			prev_hop: init_tlv_based_struct_field!(prev_hop, required),
+			value: init_tlv_based_struct_field!(value, required),
+			claim_type,
+			cltv_expiry: init_tlv_based_struct_field!(cltv_expiry, required),
+		})
+	}
+}
```

Comment:
Seems like as we change our data format more, the more complex our (de)serialization logic becomes as we can no longer use `impl_writeable_tlv_based`. I'm wondering if it would be any better to bump the version but support reading older versions. Then maybe you could condition on the version read and use a macro for implementing each version we'd like to support? I might be missing something about this, though.

Reply:
Hmm, I dont think bumping the version helps a ton, then we'd just have two copies of the deserialization logic - one for the old objects, and one for the new, and we'd have to switch between them based on the version. That just results in a code blowup that's almost equivalent, I think.

Reply:
I guess then the versions would at least be fairly well delineated. Otherwise, if we refactor further there is more cognitive overhead in figuring out the various cases. But maybe at that point we'd just decided to bump versions. I don't feel strongly on this, but just want to point out the cost.

Reply:
That said, if our version is only in topic-level structs (is it?) then we'd have to pass it through. So maybe not very feasible / clean?

Reply:
Right, I have a feeling sooner or later we'll adapt our read trait to pass a version number. I haven't bothered yet, but it may be required eventually. I agree we should be careful about growing bloat in read functions, maybe it makes sense to have a "we only support things written by versions as old as one year ago"-style policy.